src/storage/src/source/mod.rs

```diff
@@ -806,169 +866,640 @@ where
+                extended_source_upper.extend(
+                    unconsumed_partitions
+                        .into_iter()
+                        .map(|pid| (pid, MzOffset { offset: u64::MAX })),
+                );
+
+                let non_definite_errors = non_definite_errors
+                    .into_iter()
+                    .map(|e| SourceError {
+                        source_id: id,
+                        error: e.inner,
+                    })
+                    .collect_vec();
+
+                let message_batch = SourceMessageBatch {
+                    messages,
+                    non_definite_errors,
+                    source_upper: extended_source_upper,
+                };
+                // Wrap in an Rc to avoid cloning when sending it on.
+                let message_batch = Rc::new(RefCell::new(Some(message_batch)));
+
+                let cap = cap_set.delayed(&batch_counter);
+                let mut session = output.session(&cap);
+
+                session.give((message_batch, source_upper_summary));
+
+                batch_counter += 1;
+
+                if timer.elapsed() > YIELD_INTERVAL {
+                    let _ = activator.activate();
+                    break;
+                }
+            }
+        }
+    });
+
+    // TODO: Roll all this into one source operator.
+    let operator_name = format!("source_reader({})-demux", id);
+    let mut demux_op = OperatorBuilder::new(operator_name, scope.clone());
+
+    let mut input = demux_op.new_input(&stream, Pipeline);
+    let (mut batch_output, batch_stream) = demux_op.new_output();
+    let (mut summary_output, summary_stream) = demux_op.new_output();
+    let summary_output_port = summary_stream.name().port;
 
-                        cap_set.downgrade(upper.iter());
+    demux_op.build(move |_caps| {
+        let mut buffer = Vec::new();
 
-                        if upper.is_empty() {
-                            // Return early because this source is done now.
-                            break;
+        move |_frontiers| {
+            input.for_each(|cap, data| {
+                data.swap(&mut buffer);
+
+                let mut batch_output = batch_output.activate();
+                let mut summary_output = summary_output.activate();
+
+                for (message_batch, source_upper) in buffer.drain(..) {
+                    let mut session = batch_output.session(&cap);
+                    session.give(message_batch);
+
+                    let summary_cap = cap.delayed_for_output(cap.time(), summary_output_port);
+                    let mut session = summary_output.session(&summary_cap);
+                    session.give(source_upper);
+                }
+            });
+        }
+    });
+
+    ((batch_stream, summary_stream), Some(capability))
+}
+
+/// Mints new contents for the remap shard based on summaries about the source
+/// upper it receives from the raw reader operators.
```

Comment:
```suggestion
/// A _single-instance_ operator that mints new contents for the remap shard based on summaries about the source
/// upper it receives from the raw reader operators.
```

Reply:
Added something to that effect to the rustdoc but I think it's better to not use `single-instance operator` because I don't know how clear the meaning of this is to most folks.